File: backend/billirae_backend/app/db/models/invoice.py
from typing import List, Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field

class InvoiceItem(BaseModel):
    """Model for an invoice item."""
    service: str
    quantity: int
    unit_price: float
    tax_rate: float = 0.19  # Default German VAT rate

class InvoiceInDB(BaseModel):
    """Model for an invoice stored in the database."""
    id: Optional[str] = None
    user_id: str
    client_id: str
    invoice_number: str
    invoice_date: datetime = Field(default_factory=datetime.now)
    due_date: Optional[datetime] = None
    items: List[InvoiceItem]
    subtotal: float
    tax_amount: float
    total: float
    status: str = "draft"  # draft, sent, paid, overdue, cancelled
    notes: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    sent_date: Optional[datetime] = None
    paid_date: Optional[datetime] = None

    @classmethod
    async def get_by_id(cls, invoice_id: str) -> Optional['InvoiceInDB']:
        """Get invoice by ID."""
        # This would normally query the database
        # For now, return a mock invoice for testing
        return cls(
            id=invoice_id,
            user_id="user123",
            client_id="client123",
            invoice_number="INV-2023-001",
            items=[
                InvoiceItem(
                    service="Massage",
                    quantity=3,
                    unit_price=80.0,
                    tax_rate=0.19
                )
            ],
            subtotal=240.0,
            tax_amount=45.6,
            total=285.6
        )

    async def save(self) -> 'InvoiceInDB':
        """Save invoice to database."""
        self.updated_at = datetime.now()
        # This would normally save to the database
        return self

File: billirae-frontend/src/types/invoice.ts
import { SupabaseInvoice, SupabaseInvoiceItem } from '../services/supabaseClient';

export interface InvoiceItem {
  id?: string;
  invoice_id?: string;
  position_order?: number;
  service: string;
  description?: string;
  quantity: number;
  unit_price: number;
  discount?: number;
  vat: number;
  tax_note?: string;
  total?: number;
  created_at?: string;
}

export interface InvoiceData {
  id?: string;
  user_id?: string;
  client_id: string;
  date: string;
  inv_number?: string;
  notes?: string;
  status?: string;
  created_at?: string;
  items: InvoiceItem[];
  currency?: string;
  language?: string;
}

export const mapVoiceDataToInvoiceData = (voiceData: {
  client: string;
  service: string;
  quantity: number;
  unit_price: number;
  tax_rate: number;
  invoice_date: string;
  currency: string;
  language: string;
}): InvoiceData => {
  const total = voiceData.quantity * voiceData.unit_price;
  
  return {
    client_id: voiceData.client, // This will be replaced with actual client_id later
    date: voiceData.invoice_date,
    currency: voiceData.currency,
    language: voiceData.language,
    items: [
      {
        service: voiceData.service,
        quantity: voiceData.quantity,
        unit_price: voiceData.unit_price,
        vat: voiceData.tax_rate,
        total: total
      }
    ]
  };
};

export const mapSupabaseToInvoiceData = (
  invoice: SupabaseInvoice, 
  invoiceItems: SupabaseInvoiceItem[]
): InvoiceData => {
  return {
    id: invoice.id,
    user_id: invoice.user_id,
    client_id: invoice.client_id,
    date: invoice.date,
    inv_number: invoice.inv_number,
    notes: invoice.notes,
    status: invoice.status,
    created_at: invoice.created_at,
    items: invoiceItems.map(item => ({
      id: item.id,
      invoice_id: item.invoice_id,
      position_order: item.position_order,
      service: item.service,
      description: item.description,
      quantity: item.quantity,
      unit_price: item.unit_price,
      discount: item.discount,
      vat: item.vat,
      tax_note: item.tax_note,
      total: item.total,
      created_at: item.created_at
    }))
  };
};
